kratos/utilities/dof_utilities/block_build_dof_array_utility.h

```diff
@@ -0,0 +1,208 @@
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ruben Zorrilla
+//
+
+#pragma once
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/model_part.h"
+
+namespace Kratos
+{
+
+///@addtogroup KratosCore
+///@{
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @class BlockBuildDofArrayUtility
+ * @ingroup KratosCore
+ * @brief Utility class for handling the block build of DOFs
+ * @details This static class collects all the methods required to
+ * handle the DOFs when doing a block build.
+ * @author Ruben Zorrilla
+ */
+class KRATOS_API(KRATOS_CORE) BlockBuildDofArrayUtility
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of BlockBuildDofArrayUtility
+    KRATOS_CLASS_POINTER_DEFINITION(BlockBuildDofArrayUtility);
+
+    /// DOFs array type definition from ModelPart (note that this is PointerVectorSet<DofType>)
+    using DofsArrayType = ModelPart::DofsArrayType;
+
+    /// DOFs vector type definition from ModelPart (note that this is std::vector<DofType::Pointer>)s
+    using DofsVectorType = ModelPart::DofsVectorType;
+
+    /// Auxilary DOF set type definition
+    using AuxiliaryDofsSetType = std::unordered_set<Node::DofType::Pointer, DofPointerHasher>;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    BlockBuildDofArrayUtility() = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    static void SetUpDofArray(
+        const ModelPart& rModelPart,
+        DofsArrayType& rDofArray,
+        const unsigned int EchoLevel = 0,
+        const bool CheckReactionDofs = false)
+    {
+        KRATOS_TRY;
+
+        // Check that the provided DOF set is empty
+        if (rDofArray.size() != 0) {
+            KRATOS_WARNING_IF("BlockBuildDofArrayUtility", (EchoLevel > 1 && rModelPart.GetCommunicator().MyPID() == 0)) << "Provided DOF set is not empty. About to clear it." << std::endl;
+            rDofArray.clear();
+        }
+
+        // Allocate auxiliary arrays
+        DofsVectorType dof_list;
+        DofsVectorType second_dof_list; // NOTE: The second dof list is only used on constraints to include master/slave relations
+
+        AuxiliaryDofsSetType dof_global_set;
+        dof_global_set.reserve(rModelPart.NumberOfElements() * 20);
+
+        KRATOS_INFO_IF("BlockBuildDofArrayUtility", (EchoLevel > 1 && rModelPart.GetCommunicator().MyPID() == 0)) << "Setting up the DOFs" << std::endl;
+        KRATOS_INFO_IF("BlockBuildDofArrayUtility", (EchoLevel > 2 && rModelPart.GetCommunicator().MyPID() == 0)) << "Number of threads: " << ParallelUtilities::GetNumThreads() << std::endl;
+
+        // Fill the global DOF set array
+        #pragma omp parallel firstprivate(dof_list, second_dof_list)
```

Comment:
I would profit to properly refactor into new parallel utilities

Reply:
Definitively no. The intention is to avoid any change on the implementation to avoid any risk in current working implementation.

Reply:
Okay